Readme.md
# retext-syllable [![Build Status](https://img.shields.io/travis/wooorm/retext-syllable.svg?style=flat)](https://travis-ci.org/wooorm/retext-syllable) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-syllable.svg?style=flat)](https://coveralls.io/r/wooorm/retext-syllable?branch=master)

Count syllables with **[retext](https://github.com/wooorm/retext "Retext")**.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-syllable
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-syllable
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-syllable
```

[Duo](http://duojs.org/#getting-started):

```javascript
var syllable = require('wooorm/retext-syllable');
```

## Usage

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var syllable = require('retext-syllable');

var retext = new Retext()
    .use(visit)
    .use(inspect)
    .use(syllable);

retext.parse('A yellow fresh banana.', function (err, tree) {
    /* Log the first sentence and its words, note its data: */
    console.log(tree.head.head);

    /**
     * SentenceNode[8] [data={"syllableCount":7}]
     * ├─ WordNode[1] [data={"syllableCount":1}]
     * │  └─ TextNode: 'A'
     * ├─ WhiteSpaceNode: ' '
     * ├─ WordNode[1] [data={"syllableCount":2}]
     * │  └─ TextNode: 'yellow'
     * ├─ WhiteSpaceNode: ' '
     * ├─ WordNode[1] [data={"syllableCount":1}]
     * │  └─ TextNode: 'fresh'
     * ├─ WhiteSpaceNode: ' '
     * ├─ WordNode[1] [data={"syllableCount":3}]
     * │  └─ TextNode: 'banana'
     * └─ PunctuationNode: '.'
     */
});
```

## API

None, **retext-syllable** automatically detects the syllable count of each [`WordNode`](https://github.com/wooorm/textom/tree/master#textomwordnode-nlcstwordnode) (using **[wooorm/syllable](https://github.com/wooorm/syllable)**), and stores the count in `node.data.syllableCount`.

Additionally, all parents (such as sentences, paragraphs, root) also receive a `syllableCount` data property (`parent.data.syllableCount`).

## Performance

On a MacBook Air, **retext** performs about 40% slower with **retext-syllable**.

```text
           retext w/o retext-syllable
  225 op/s » A paragraph (5 sentences, 100 words)
   25 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)

           retext w/ retext-syllable
  136 op/s » A paragraph (5 sentences, 100 words)
   14 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.1.6 / 2015-01-19
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Refactor to use generic `changeinside` in new retext
  * Update retext to stable release

0.1.5 / 2014-12-08
==================

 * Add benchmark
 * Fix wording in `Readme.md`
 * Refactor npm script targets in `package.json`
 * Add link to personal website to copyright in `Readme.md`
 * Update eslint, retext

0.1.4 / 2014-11-21
==================

 * Refactor `Readme.md`
 * Add flat badges to `Readme.md`
 * Fix incorrect executive rights on `test.js`
 * Refactor npm script targets in `package.json`
 * Add retext-inspect to dev-dependecies
 * Update retext

0.1.3 / 2014-10-28
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext, retext-content, retext-visit

0.1.2 / 2014-10-21
==================

 * Refactor .jscs.json
 * Update retext

0.1.1 / 2014-10-15
==================

 * Update tests for changes in retext-visit
 * Update syllable, retext, retext-content, retext-visit
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js

0.1.0 / 2014-09-27
==================

 * Refactor Readme.md
 * Refactor API
 * Remove browser test
 * Update .gitignore, .npmignore
 * Fix property order in component.json, package.json, bower.json
 * Refactor spec
 * Update retext, retext-visit, retext-content
 * Fix several typo’s


test.js
'use strict';

/*
 * Dependencies.
 */

var syllable,
    content,
    visit,
    inspect,
    Retext,
    assert;

syllable = require('./');
content = require('retext-content');
Retext = require('retext');
visit = require('retext-visit');
inspect = require('retext-inspect');
assert = require('assert');

/*
 * Retext.
 */

var retext;

retext = new Retext()
    .use(visit)
    .use(inspect)
    .use(content)
    .use(syllable);

/*
 * Fixtures.
 */

var sentence,
    counts,
    parentCount,
    otherWords,
    otherCounts,
    otherParentCount;

sentence = 'A simple, english, sentence';
counts = [1, 2, 2, 2];
otherWords = ['A', 'yellow', 'fresh', 'banana'];
otherCounts = [1, 2, 1, 3];
parentCount = 7;
otherParentCount = 7;

/*
 * Tests.
 */

describe('syllable()', function () {
    var tree;

    before(function (done) {
        retext.parse(sentence, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof syllable === 'function');
    });

    it('should count each `WordNode`', function () {
        var index;

        index = -1;

        tree.visit(tree.WORD_NODE, function (node) {
            index++;

            assert('syllableCount' in node.data);
            assert(node.data.syllableCount === counts[index]);
        });
    });

    it('should count `syllableCount` in parents', function () {
        assert(tree.data.syllableCount === parentCount);
        assert(tree.head.data.syllableCount === parentCount);
        assert(tree.head.head.data.syllableCount === parentCount);
    });

    it('should set each `syllableCount` to `0` when a `WordNode` has no ' +
        'value',
        function () {
            tree.visit(tree.WORD_NODE, function (node) {
                node.removeContent();

                assert(node.data.syllableCount === 0);
            });
        }
    );

    it('should reset `syllableCount` in parents', function () {
        assert(tree.data.syllableCount === 0);
        assert(tree.head.data.syllableCount === 0);
        assert(tree.head.head.data.syllableCount === 0);
    });

    it('should re-count `WordNode`s when their values change', function () {
        var index;

        index = -1;

        tree.visit(tree.WORD_NODE, function (node) {
            index++;

            node.replaceContent(otherWords[index]);

            assert(node.data.syllableCount === otherCounts[index]);
        });
    });

    it('should re-count `syllableCount` in parents', function () {
        assert(tree.data.syllableCount === otherParentCount);
        assert(tree.head.data.syllableCount === otherParentCount);
        assert(tree.head.head.data.syllableCount === otherParentCount);
    });
});


spec/retext-syllable.spec.js
'use strict';

/**
 * Dependencies.
 */

var syllable,
    content,
    visit,
    Retext,
    assert;

syllable = require('..');
content = require('retext-content');
Retext = require('retext');
visit = require('retext-visit');
assert = require('assert');

/**
 * Retext.
 */

var retext;

retext = new Retext()
    .use(visit)
    .use(content)
    .use(syllable);

/**
 * Fixtures.
 */

var sentence,
    counts,
    parentCount,
    otherWords,
    otherCounts,
    otherParentCount;

sentence = 'A simple, english, sentence';
counts = [1, 2, 2, 2];
otherWords = ['A', 'yellow', 'fresh', 'banana'];
otherCounts = [1, 2, 1, 3];
parentCount = 7;
otherParentCount = 7;

/**
 * Tests.
 */

describe('syllable()', function () {
    var tree;

    before(function (done) {
        retext.parse(sentence, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should be a `function`', function () {
        assert(typeof syllable === 'function');
    });

    it('should export an `attach` method', function () {
        assert(typeof syllable.attach === 'function');
    });

    it('should count each `WordNode`', function () {
        var index;

        index = -1;

        tree.visitType(tree.WORD_NODE, function (node) {
            index++;

            assert('syllableCount' in node.data);
            assert(node.data.syllableCount === counts[index]);
        });
    });

    it('should count `syllableCount` in parents', function () {
        assert(tree.data.syllableCount === parentCount);
        assert(tree.head.data.syllableCount === parentCount);
        assert(tree.head.head.data.syllableCount === parentCount);
    });

    it('should set each `syllableCount` to `0` when a `WordNode` has no ' +
        'value',
        function () {
            tree.visitType(tree.WORD_NODE, function (node) {
                node.removeContent();

                assert(node.data.syllableCount === 0);
            });
        }
    );

    it('should reset `syllableCount` in parents', function () {
        assert(tree.data.syllableCount === 0);
        assert(tree.head.data.syllableCount === 0);
        assert(tree.head.head.data.syllableCount === 0);
    });

    it('should re-count `WordNode`s when their values change', function () {
        var index;

        index = -1;

        tree.visitType(tree.WORD_NODE, function (node) {
            index++;

            node.replaceContent(otherWords[index]);

            assert(node.data.syllableCount === otherCounts[index]);
        });
    });

    it('should re-count `syllableCount` in parents', function () {
        assert(tree.data.syllableCount === otherParentCount);
        assert(tree.head.data.syllableCount === otherParentCount);
        assert(tree.head.head.data.syllableCount === otherParentCount);
    });
});


benchmark.js
'use strict';

var Retext,
    syllable;

/*
 * Dependencies.
 */

Retext = require('retext');
syllable = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithSyllable;

retext = new Retext();
retextWithSyllable = new Retext().use(syllable);

/*
 * Test data: A (big?) article (w/ 100 paragraphs, 500
 * sentences, 10,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

suite('retext w/o retext-syllable', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retext.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );
});

suite('retext w/ retext-syllable', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retextWithSyllable.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retextWithSyllable.parse(section, done);
        }
    );
});
